### PR TITLE
[Bug Fixes] fix cache_miss for first request prefill when tp_size = 1

### PR DESCRIPTION
**Pull request by aolemila**
# Bug
#233 

**Reason**
The kv cache in layers_kv_fused has lost the PartitionSpec(spec=()) when tp_size is 1. Expected sharding is spec=(None, "tensor").

**Solution**
In order to avoid influence performance for tp_size > 1 as much as possible, here add device_put for tp_size = 1.

Note: If you had better solution, please comment it. Thank you!

**Result**
Note: The launch and curl commands are used in #233.
```bash
[2025-10-11 11:16:45] Prefill batch. #new-seq: 1, #new-token: 128, #cached-token: 0, token usage: 0.00, #running-req: 0, #queue-req: 0,
[2025-10-11 11:16:45] INFO:     127.0.0.1:43020 - "POST /v1/chat/completions HTTP/1.1" 200 OK
```

**Other Work**
1. Add cache_miss test codes for tp_size  = 1

# Command
```bash
# case1: tp_size=1; case2: tp_size = 4
JAX_COMPILATION_CACHE_DIR=/tmp/jit_cache \
python3 -u -m sgl_jax.launch_server \
--model-path Qwen/Qwen3-8B \
--trust-remote-code  \
--tp-size=4 \
--device=tpu \
--mem-fraction-static=0.8 \
--chunked-prefill-size=2048 \
--download-dir=/tmp \
--dtype=bfloat16 \
--max-running-requests 256 \
--skip-server-warmup \
--page-size=128  \
--disable-radix-cache

evalscope eval  --model Qwen-8B --api-url http://127.0.0.1:30000/v1/chat/completions --api-key EMPTY --eval-type service --datasets gsm8k --eval-batch-size 64

python3 -m sgl_jax.bench_serving --backend sgl-jax --dataset-name random --num-prompts 48 --random-input 1024 --random-output 1024 --random-range-ratio 1 --warmup-requests 0 --max-concurrency=16
```

# Benchmark

### current banch

```bash
============ Serving Benchmark Result ============
Backend:                                 sgl-jax
Traffic request rate:                    inf
Max request concurrency:                 16
Successful requests:                     48
Benchmark duration (s):                  55.78
Total input tokens:                      49152
Total generated tokens:                  49152
Total generated tokens (retokenized):    49141
Request throughput (req/s):              0.86
Input token throughput (tok/s):          881.22
Output token throughput (tok/s):         881.22
Total token throughput (tok/s):          1762.44
Concurrency:                             15.99
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   18584.82
Median E2E Latency (ms):                 18580.35
---------------Time to First Token----------------
Mean TTFT (ms):                          311.73
Median TTFT (ms):                        317.28
P99 TTFT (ms):                           520.58
---------------Inter-Token Latency----------------
Mean ITL (ms):                           17.86
Median ITL (ms):                         17.66
P95 ITL (ms):                            18.77
P99 ITL (ms):                            19.11
Max ITL (ms):                            481.18
==================================================
```

### main

```bash
============ Serving Benchmark Result ============
Backend:                                 sgl-jax
Traffic request rate:                    inf
Max request concurrency:                 16
Successful requests:                     48
Benchmark duration (s):                  56.55
Total input tokens:                      49152
Total generated tokens:                  49152
Total generated tokens (retokenized):    49141
Request throughput (req/s):              0.85
Input token throughput (tok/s):          869.23
Output token throughput (tok/s):         869.23
Total token throughput (tok/s):          1738.46
Concurrency:                             15.99
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   18840.96
Median E2E Latency (ms):                 18610.41
---------------Time to First Token----------------
Mean TTFT (ms):                          328.13
Median TTFT (ms):                        316.84
P99 TTFT (ms):                           930.36
---------------Inter-Token Latency----------------
Mean ITL (ms):                           18.10
Median ITL (ms):                         17.66
P95 ITL (ms):                            18.77
P99 ITL (ms):                            19.15
Max ITL (ms):                            1224.85
==================================================
```

## case2: tp_size = 4

```bash
============ Serving Benchmark Result ============
Backend:                                 sgl-jax
Traffic request rate:                    inf
Max request concurrency:                 16
Successful requests:                     48
Benchmark duration (s):                  19.08
Total input tokens:                      49152
Total generated tokens:                  49152
Total generated tokens (retokenized):    49113
Request throughput (req/s):              2.52
Input token throughput (tok/s):          2575.77
Output token throughput (tok/s):         2575.77
Total token throughput (tok/s):          5151.54
Concurrency:                             15.99
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   6356.72
Median E2E Latency (ms):                 6359.96
---------------Time to First Token----------------
Mean TTFT (ms):                          155.96
Median TTFT (ms):                        161.62
P99 TTFT (ms):                           261.41
---------------Inter-Token Latency----------------
Mean ITL (ms):                           6.06
Median ITL (ms):                         5.95
P95 ITL (ms):                            6.21
P99 ITL (ms):                            6.37
Max ITL (ms):                            227.48
==================================================
```

# Accuracy

## case1: tp_size = 1

```bash
+---------+-----------+-----------------+----------+-------+---------+---------+
| Model   | Dataset   | Metric          | Subset   |   Num |   Score | Cat.0   |
+=========+===========+=================+==========+=======+=========+=========+
| Qwen-8B | gsm8k     | AverageAccuracy | main     |  1319 |  0.9151 | default |
+---------+-----------+-----------------+----------+-------+---------+---------+
```

## case2: tp_size = 4

```bash
+---------+-----------+-----------------+----------+-------+---------+---------+
| Model   | Dataset   | Metric          | Subset   |   Num |   Score | Cat.0   |
+=========+===========+=================+==========+=======+=========+=========+
| Qwen-8B | gsm8k     | AverageAccuracy | main     |  1319 |  0.9083 | default |
+---------+-----------+-----------------+----------+-------+---------+---------+
```

### CODE REVIEW

File: python/sgl_jax/srt/model_executor/model_runner.py [hunk 1]
```diff
@@ -395,6 +395,17 @@ def _forward(
...
+            self.token_to_kv_pool.mesh, P(None, self.token_to_kv_pool.kv_partition_axis)
+        )
+        # note: for tp_size == 1, we need to put the layers_kv_fused on the device with the target_sharding
+        # because sharding P(None, 'tensor') constraint has lost and this results in cache_miss for first prefill phase.
+        # issue: https://github.com/sgl-project/sglang-jax/issues/233
+        if self.tp_size == 1:
+            layers_kv_fused = [
+                jax.device_put(layer_kv_fused, target_sharding)
+                for layer_kv_fused in layers_kv_fused
+            ]
```
[the remark this answers is not in the document]
Comment (aolemila) [reply python/sgl_jax/srt/model_executor/model_runner.py]: done
